include/deal.II/base/function_spherical.h

```diff
@@ -0,0 +1,100 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__function_spherical_h
+#define dealii__function_spherical_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/function.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/std_cxx11/array.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Functions
+{
+
+  /**
+   * A base class scalar-valued function $f=f(r,\theta,\phi)$ defined in
+   * spherical coordinates. The convention for angles is the same as in
+   * GeometricUtilities::Coordinates.
+   *
+   * This is an abstract class which wraps transformation of values, gradients
+   * and hessians from spherical coordinates to Cartesian coordinate system.
+   * Therefore derived classes only need to implement those functions in
+   * spherical coordiantes (see svalue(), sgradient() and
+   * shessian() ).
+   *
+   * @note This function is currently only implemented for dim==3 .
+   *
+   * @author Denis Davydov
+   * @date 2017
+   */
+  template <int dim>
+  class Spherical : public Function<dim>
+  {
+  public:
+    /**
+     * Constructor which should be provided with center @p center.
+     */
+    Spherical(const Point<dim> &center);
+
+    /**
+     * Virtual destructor.
+     */
+    virtual ~Spherical();
+
+    virtual double value (const Point<dim> &point,
+                          const unsigned int component = 0) const;
+
+    virtual Tensor<1,dim> gradient (const Point<dim>   &p,
+                                    const unsigned int  component = 0) const;
+
+    virtual SymmetricTensor<2,dim> hessian (const Point<dim> &p,
+                                            const unsigned int component=0) const;
+
+    std::size_t memory_consumption () const;
+
+  private:
+    /**
+     * Value in spherical coordinates
+     */
+    virtual double svalue(const std_cxx11::array<double, dim> &sp) const = 0;
+
+    /**
+     * Gradient in spherical coordinates
+     */
+    virtual Tensor<1,dim> sgradient(const std_cxx11::array<double, dim> &sp) const = 0;
```

Comment:
I've really taken to think that just because something is *convenient* to represent one way doesn't necessarily make it *useful* in the long run. In particular, I have over the years really come to think that *readability* of code is a really important thing, and that means that we should use `Tensor` only for things that really *are* tensors.

In your case, I would suggest using `array` for the return type, or to invent a new class.